src/psyclone/psyir/backend/fortran.py

```diff
@@ -0,0 +1,443 @@
+        been addressed this module can be added only when required.
+
+        :param node: A KernelSchedule PSyIR node.
+        :type node: :py:class:`psyclone.psyGen.KernelSchedule`
+
+        :returns: The Fortran code as a string.
+        :rtype: str
+
+        '''
+        if not node.name:
+            raise VisitorError("Expected node name to have a value.")
+
+        result = (
+            "{0}module {1}\n"
+            "".format(self._nindent, node.name+"_mod"))
+
+        self._depth += 1
+        args = [symbol.name for symbol in node.symbol_table.argument_list]
+        result += (
+            "{0}use constants_mod, only : r_def, i_def\n"
+            "{0}implicit none\n"
+            "{0}contains\n"
+            "{0}subroutine {1}({2})\n"
+            "".format(self._nindent, node.name, ",".join(args)))
+
+        self._depth += 1
+        # Declare the kernel data.
+        declarations = ""
+        for symbol in node.symbol_table._symbols.values():
+            declarations += self.get_declaration(symbol)
+        # Get the executable statements.
+        exec_statements = ""
+        for child in node.children:
+            exec_statements += self.visit(child)
+        result += (
+            "{0}\n"
+            "{1}\n"
+            "".format(declarations, exec_statements))
+
+        self._depth -= 1
+        result += (
+            "{0}end subroutine {1}\n"
+            "".format(self._nindent, node.name))
+
+        self._depth -= 1
+        result += (
+            "{0}end module {1}\n"
+            "".format(self._nindent, node.name+"_mod"))
+        return result
+
+    def assignment(self, node):
+        '''This method is called when an Assignment instance is found in the
+        PSyIR tree.
+
+        :param node: An Assignment PSyIR node.
+        :type node: :py:class:`psyclone.psyGen.Assigment`
+
+        :returns: The Fortran code as a string.
+        :rtype: str
+
+        '''
+        lhs = self.visit(node.children[0])
+        rhs = self.visit(node.children[1])
+        result = "{0}{1}={2}\n".format(self._nindent, lhs, rhs)
+        return result
+
+    def binaryoperation(self, node):
+        '''This method is called when a BinaryOperation instance is found in
+        the PSyIR tree.
+
+        :param node: A BinaryOperation PSyIR node.
+        :type node: :py:class:`psyclone.psyGen.BinaryOperation`
+
+        :returns: The Fortran code as a string.
+        :rtype: str
+
+        '''
+        # reverse the fortran2psyir mapping to make a psyir2fortran
+        # mapping
+        from psyclone.psyGen import Fparser2ASTProcessor as f2psyir
+        mapping = {}
+        for operator in f2psyir.binary_operators:
+            mapping_key = f2psyir.binary_operators[operator]
+            mapping_value = operator
+            # Only choose the first mapping value when there is more
+            # than one.
+            if mapping_key not in mapping:
+                mapping[mapping_key] = mapping_value
```

Comment:
Nice one. 

Not for this PR but I had a similar idea for this front-end<->back-end duplications by using Bidirectional HashTables (https://stackoverflow.com/a/21894086) which is basically what you did, but I think they can be created just once and then just fetched from the front-end and back-end.

Reply:
That would be great. The only thing I would like is to be able to consistently choose particular values when there is more than one choice e.g. `!=` and `.ne.`. We could just use sort as Andy suggested in earlier discussions but I think it would be better to be able to choose a class of choices consistently e.g. `==`, `<=`, `!=` vs `.eq.`, `.le.`, `.ne.`.